### PR TITLE
Whitelist feature

### PR DESCRIPTION
**Pull request by denis-y-krasilnikov**
Helpers have been splitted to common, case and describe.Version support has been added to describe strategy. Format [KEY (VERSION)].
Whitelist feature has been implemented: test case marks as blocked if it isn't in whitelist or versions aren't the same.
--noUpdateCases flag has been added: disable updating test case according to result steps.
Steps have been disabled for describe strategy.
Failure messages have been added as a comment for describe strategy.

### CODE REVIEW

File: lib/cli/index.js [hunk 1]
```diff
@@ -11,19 +11,18 @@ const reports = fs.readJsonSync(config.reportsFile.JSON);
...
+    const { testRunKey, target, noUpdateCases } = opts;
     const testHandler = factory.create(target);
 
     if (testHandler === null) throw Error('Could not determine testHandler');
     const testResults = testHandler.getTestResults(reports);
-    const result = testHandler.formingScripts(testResults);
+    const result = await testHandler.formingScripts(testResults);
     const resultTestCase = testHandler.formingSteps(testResults);
 
+    if (!noUpdateCases) await testCase.prepare(resultTestCase);
```
Comment (violgadk) [inline lib/cli/index.js]: maybe rename? has... is...?

File: lib/test-handler/helpers/describe.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    formingSteps,
+} = require('./common');
+
+/**
+ * Separate test key to key and version.
+ *
+ * @param {string} testKey - Test key.
+ * @returns {{version: string, key: string}} Separated key and version.
+ */
+const getTestKeyVersion = testKey => {
```
Comment (violgadk) [inline lib/test-handler/helpers/describe.js]: strange name

File: lib/test-handler/helpers/case.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    return resultsList;
+};
+
+module.exports = {
+    reportException,
+    getTestKey,
+    formatStatus,
+    getRelativePath,
+    formingScripts,
+    formingSteps,
```
Comment (violgadk) [inline lib/test-handler/helpers/case.js]: getSteps? getScripts? may be...

File: lib/test-handler/helpers/describe.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+ */
+const getScriptErrors = testResult => testResult.map(testCase => testCase.comment).join('\n');
+
+const getTestCaseWhitelist = async () =>
+    (await api.testCase.search(config.whitelist.query || '')).reduce(
+        (res, it) => ({ ...res, [it.key]: it.majorVersion }),
+        {},
+    );
+
+/**
```
Comment (Alexandrshy) [inline lib/test-handler/helpers/describe.js]: Remove it

File: lib/test-handler/helpers/describe.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+ * Get script errors.
+ *
+ * @param {Array} testResult - Result of the test case.
+ * @returns {Object} Script errors for JTM.
+ */
+const getScriptErrors = testResult => testResult.map(testCase => testCase.comment).join('\n');
+
+const getTestCaseWhitelist = async () =>
```
Comment (Alexandrshy) [inline lib/test-handler/helpers/describe.js]: What about the description in JSDoc of this feature?

File: lib/test-handler/helpers/describe.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    (await api.testCase.search(config.whitelist.query || '')).reduce(
+        (res, it) => ({ ...res, [it.key]: it.majorVersion }),
```
Comment (Alexandrshy) [inline lib/test-handler/helpers/describe.js]: I think it should be a separate method with a description of what is happening here

File: lib/config/config-schema.json [hunk 7]
```diff
@@ -45,6 +45,19 @@
             "required": [
                 "host"
             ]
+        },
+        "whitelist": {
```
Comment (Alexandrshy) [inline lib/config/config-schema.json]: I need to describe this in the README

**Comment by Alexandrshy** [comment]
Each change must be recorded in CHANGELOG.md with a description of the changes. You also need to change the version

File: lib/cli/index.js (hunk 1 above)
Comment (Alexandrshy) [inline lib/cli/index.js]: I think it needs to be changed
I think we should use the option to update the cases, we should not use negation

File: lib/test-handler/helpers/describe.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+} = require('./common');
+
+/**
+ * Separate test key to key and version.
+ *
+ * @param {string} testKey - Test key.
+ * @returns {{version: string, key: string}} Separated key and version.
+ */
+const getTestKeyVersion = testKey => {
+    const result = RegExp(/(.*?)(?:\((.*)\))/).exec(testKey);
```
Comment (Alexandrshy) [inline lib/test-handler/helpers/describe.js]: You also need to describe an example of the name of the test with the key and version in README

Comment (Alexandrshy) [reply lib/test-handler/helpers/describe.js]: And as I see you changed the template to describe the key, what was it done for? I think to add a version you had to expand the existing template.

File: lib/config/config-schema.json (hunk 7 above)
Comment (Alexandrshy) [reply lib/config/config-schema.json]: In general, why do you use a config file for this? This should work like this:
1. We executed the tcm-export: case command and got the result in TCM for the report with the key in "case";
2. We executed the tcm-export: describe command - and got the result in TCM for the report with the key in the "describe"
3. We ran the tcm-export: describeversio command (this is just an example), it is "node ./bin --target describe --cycleName testdecribe --checkVersion" (and this is also an example) and already the option --checkVersion responsible for checking the version in the test

File: lib/test-handler/helpers/describe.js (hunk 6 above)
Comment (denis-y-krasilnikov) [reply lib/test-handler/helpers/describe.js]: i've ranamed veriables to make it more readable

File: lib/test-handler/helpers/describe.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+ */
+const getTestCaseWhitelist = async () =>
+    (await api.testCase.search(config.whitelist.query || '')).reduce(
+        (whitelist, testCase) => ({ ...whitelist, [testCase.key]: testCase.majorVersion }),
+        {},
+    );
+
+/**
+ * Format result for Test Cycles.
+ *
```
Comment (CptHeartless) [inline lib/test-handler/helpers/describe.js]: Describe please, from which and to which this function does format

File: lib/test-handler/describe.js [hunk 10]
```diff
@@ -20,31 +20,30 @@ const {
...
+    const name = getTestName(dataResult);
+    const key = getTestKey(name);
+
+    if (!key) {
+        reportException(name);
+        return {};
+    }
 
     return testResults.reduce((accumulator, test) => {
-        const { title, status, duration } = test;
```
Comment (CptHeartless) [inline lib/test-handler/describe.js]: why you removed "duration"?

File: lib/test-handler/helpers/describe.js (hunk 9 above)
Comment (denis-y-krasilnikov) [reply lib/test-handler/helpers/describe.js]: it works the same as for case strategy, but also marks tests as blocked if they are not in whitelist

File: lib/test-handler/describe.js (hunk 10 above)
Comment (denis-y-krasilnikov) [reply lib/test-handler/describe.js]: actually i don't remember, but i think that it's useless or not applicable for describe strategy

**Pull request merged**